Add input_text component

BioSehnsucht opened the pull request:
Relates to home-assistant/home-assistant#9112

UI side of input_text component implementation.

BioSehnsucht commented:
I see that Travis CI failed. Is this due to not having implemented tests yet, or something else? The output doesn't seem to indicate anything obvious to me (this is my first go at Polymer stuff)

BioSehnsucht committed:
Add input_text component

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,92 @@
+</dom-module>
+
+<script>
+Polymer({
```
I know that you probably copy pasted this from other components, but we're in the process of migrating to Polymer 2 classes for components. We should no longer add Polymer 1 components.

See [this component](https://github.com/home-assistant/home-assistant-polymer/blob/206f624fb6008f19ae1c514a16ce77461a9cfbcd/src/state-summary/state-card-content.html) for an example of a Polymer 2 class. The conversion is very straight forward.

BioSehnsucht replied:
Yes, it started as a copy of input_slider. I'll figure out the changes and I guess do the same to input_slider later, since it should be nearly identical changes to make, when I do the rename to input_number.

BioSehnsucht commented:
@balloob Let me know if I converted this right, it appears to function identically for me. I'm not entirely sure on the property -> stateObj -> observer bit since the Polymer 2 code seems to do observer via a function, but wasn't sure if I should try to nest more class-y code in there or just leave that observer as-is. Moving the observer to the class level (outside properties) doesn't work, since it needs to observe the property I guess? There's probably a better way to do this.

BioSehnsucht committed:
Update input_text to Polymer 2

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+      <state-info state-obj="[[stateObj]]" in-dialog='[[inDialog]]'></state-info>
+      <paper-input
+        minlength='[[min]]' maxlength='[[max]]' value='{{value}}' pin
+        disabled='[[disabled]]' auto-validate pattern='[[pattern]]'
```
We should only `auto-validate` if a pattern is given?

BioSehnsucht replied:
If no pattern is provided, pattern gets set to `'.*'` by default - I suppose the alternative is to replace the whole `auto-validate pattern=....` with a `[[some string property]]` which is either empty or contains the whole `auto-validate pattern=...` but this seemed easier?

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+    <div class='horizontal justified layout'>
+      <state-info state-obj="[[stateObj]]" in-dialog='[[inDialog]]'></state-info>
+      <paper-input
+        minlength='[[min]]' maxlength='[[max]]' value='{{value}}' pin
```
`pin` ?

balloob replied:
There is no need to make `min` a variable, you can just refer to `[[stateObj.attributes.min]]`

BioSehnsucht replied:
This was created by using input_slider as a template, and I overlooked that. I'll remove `pin` since it's unnecessary / useless.

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+    if (this.value === String(this.stateObj.state)) {
+      return;
+    }
+    this.hass.callService('input_text', 'select_value', {
```
Is the service called `select_value` ? If it is, that is wrong, it should be called `set_value`.

BioSehnsucht replied:
Again, used input_slider as a starting point so that's where that came from. It's called select_value in the Python code, but I can change this to set_value in both places.

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  stateObjectChanged(newVal) {
+    this.min = Number(newVal.attributes.min);
```
No need to cast to number, they should be numbers coming in from HA. If that's not the case, we should update the component.

BioSehnsucht replied:
I'll remove these redundant casts.

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+  stateObjectChanged(newVal) {
+    this.min = Number(newVal.attributes.min);
+    this.max = Number(newVal.attributes.max);
+    if (newVal.attributes.pattern != null) {
```
If the new pattern is null, you keep using the pattern of the old state object.

BioSehnsucht replied:
Hadn't considered that pattern might be changed dynamically. I guess I'll add an else clause to (re)set it to `'.*'` (the default) when null.

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,108 @@
+    if (newVal.attributes.pattern != null) {
+      this.pattern = String(newVal.attributes.pattern);
+    }
+    this.disabled = Boolean(newVal.attributes.disabled);
```
This is not possible?

BioSehnsucht replied:
What's not possible? I've tested this, even changed the disabled state using the developer tools / states screen in HASS and had it reflect as expected (when disabled, the input text box is greyed out but still reflects changes to the value, so automations / API can still affect it just not the user via UI).

BioSehnsucht replied:
On further testing, it seems that I can set a value after changing `disabled = false` via tools / states but then it resets to `disabled = true` afterwards, if that is what was configured in the configuration. I've tried ensuring there's a setter for the disabled property on the component, but it still resets.

BioSehnsucht committed:
input_text : changes suggested in home-assistant/home-assistant-polymer#408

BioSehnsucht commented:
@balloob Made most of those changes you suggested, or at least addressed the issues raised if not changed in precisely that fashion. Still unclear on the comment about disabled being impossible, but it seems while I can set it via states (and in theory also API) to true/false, and reflect that in the UI, if it was disabled in config then after I set a new value it re-disables itself. I'm not sure why, but maybe related to your comment about it being impossible?

To change from `select_value` to `set_value` required changes in component, I haven't opened a new PR (as the original PR was merged and closed) yet but you can see the linked commit to my repo above. Didn't want to open a new PR just yet until I'm sure I'm not making any more changes on the component side :)

balloob commented:
So we will need some changes to the component which then should be reflected here:

 - change service to be set_value
 - remove disabled

balloob commented:
Added a PR with my proposed changes https://github.com/home-assistant/home-assistant/pull/9326

BioSehnsucht commented:
@balloob Is there a reason we can't have the UI gadgets be disabled from user control, but still be updated via API/automation (so that potentially an API call / automation could change the state based on some other condition)? If the feature has to go, it has to go, but I am curious as to the why.

BioSehnsucht committed:
input_text : removing disabled option per home-assistant/home-assistant-polymer#408

BioSehnsucht commented:
@balloob  I went ahead and pushed removal of disabled from Polymer, and also created a PR on github.io repo for corresponding change to documentation.

BioSehnsucht commented:
@balloob  Should probably add this to the 0.53 milestone to make sure it makes it into 0.53?

balloob commented on src/state-summary/state-card-input_text.html:
```diff
@@ -0,0 +1,91 @@
+      <state-info state-obj="[[stateObj]]" in-dialog='[[inDialog]]'></state-info>
+      <paper-input
+        minlength='[[stateObj.attributes.min]]' maxlength='[[stateObj.attributes.max]]' value='{{value}}'
+        [[pattern]]
```
This is not how the template syntax works.

balloob commented on src/state-summary/state-card-content.html:
```diff
@@ -6,6 +6,7 @@
 <link rel="import" href="state-card-display.html">
```
Please revert file permission change.

balloob replied:
Fixed it.

balloob committed:
Fix file permissions

balloob committed:
UI tweaks + cleanup

balloob commented:
Made sure it is part of 0.53
